subsys/bootloader/mcuboot_image_header_parser/mcuboot_image_header_parser.c

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr/types.h>
+#include <string.h>
+#include <image.h>
+#include <errno.h>
+
+/**@brief Reads MCUBoots Image header
+ *
+ * @param header_addr Address of where the image header is located.
+ * @param[out] image_header Pointer to the image header struct which should be
+ * populated.
+ *
+ * @retval 0 If successfully read the IMAGE_HEADER magic meaning an MCUBoot
+ * header was present.
+ * @retval -EINVAL If the image_header magic doesn't match meaning an MCUBoot
+ * header was not found at the given address.
+ */
+static int mcuboot_read_img_hdr(u32_t *header_addr, struct image_header *image_header)
+{
+	memcpy(image_header, header_addr, IMAGE_HEADER_SIZE);
+	if (image_header->ih_magic != IMAGE_MAGIC) {
+		return -EINVAL;
+	}
+	return 0;
+}
+
+uint32_t mcuboot_read_version_number(u32_t *header_addr)
```

Comment:
u32_t

Reply:
oversight (fixed.)

Reply:
To be blunt, can we just replace this file with the following?

    const struct image_version *mcuboot_get_version(u32_t header_addr)
    {
        const struct image_header * ih = (const struct image_header *)header_addr;
        return ih->ih_magic == IMAGE_MAGIC ? &ih->ih_ver : NULL;
    }

Benefits:
 - Doesn't define a new version number format. If the format is not new in our codebase, disregard this point.
 - Doesn't copy.
 - Doesn't define a new "error code" (UINT32_MAX)
 - Doesn't need a new set of files. (Maybe it could be added as a helper function in an MCUboot file, I haven't investigated).

Reply:
I'm not against what you are suggestion, we would have to add the MCUBoot header include into the b0 sample but I guess it's not really the worst thing in the world.

Also the creation of some macro which does `MCUBOOT_IMG_VER_COMPARE` which implements the same way of comparing the version numbers which is what the version number returned is supposed to be used for so it's more a data-conversion than reading.

My biggest concern with this solution is returning `NULL`. The reason I chose `UINT32_MAX` is it is less likely for people to version their first version 255.255.65535 than 0.0.0. As `&ih->ih_ver` will resolve to `NULL` if `major, minor, rev` is 0

Reply:
Here's MCUboot's own version comparison function (in case you don't know about it):
https://github.com/JuulLabs-OSS/mcuboot/blob/fe5ab1c35b12ce1545481766906275ec96d91daa/boot/bootutil/src/bootutil_misc.c#L731

Reply:
Doesn't compare which is highest.

Reply:
ver1 >= ver2: `boot_is_version_sufficient(ver2, ver1)`
ver1 > ver2: `!boot_is_version_sufficient(ver1, ver2)`

Reply:
Still doesn't help me as I can't compile the code module it is in without pulling in a lot of other stuff due to `flash_area_close` etc. So I think it would be easier to use a macro comparing them as 32-bit unsigned integers IMO

Reply:
Something more like this you wanted right? I'm a bit unsure of the comparison logic in the end and the error codes used.

Reply:
Just one question wouldn't it be more efficient to do the u32 compare instead of all the logic checks done in the mentioned mcuboot function?